src/Type/Php/ArrayReplaceFunctionReturnTypeExtension.php

```diff
@@ -24,68 +24,57 @@ public function isFunctionSupported(FunctionReflection $functionReflection): boo
 
 	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
 	{
-		$args = $functionCall->getArgs();
+		$arrayTypes = $this->collectArrayTypes($functionCall, $scope);
 
-		if (count($args) < 1) {
+		if (count($arrayTypes) === 0) {
 			return ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
 		}
 
-		$arrayType = $scope->getType($args[0]->value);
-		if ($arrayType->isArray()->yes()) {
-			$resultType = $this->getResultType($functionCall, $scope);
-
-			if ($resultType !== null) {
-				if ($this->returnsNonEmptyArray($functionCall, $scope)) {
-					$resultType = TypeCombinator::intersect($resultType, new NonEmptyArrayType());
-				}
-
-				return $resultType;
-			}
-		}
-
-		return ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+		return $this->getResultType(...$arrayTypes);
 	}
 
-	private function getResultType(FuncCall $functionCall, Scope $scope): ?Type
+	private function getResultType(Type ...$arrayTypes): Type
 	{
-		$args = $functionCall->getArgs();
-
 		$keyTypes = [];
 		$valueTypes = [];
-		for ($i = 0; $i < count($args); $i++) {
-			$arrayType = $scope->getType($args[$i]->value);
-
-			if (!$arrayType->isArray()->yes()) {
-				return null;
-			}
-
+		$nonEmptyArray = false;
+		foreach ($arrayTypes as $arrayType) {
 			$keyTypes[] = $arrayType->getIterableKeyType();
 			$valueTypes[] = $arrayType->getIterableValueType();
+			if ($nonEmptyArray || !$arrayType->isIterableAtLeastOnce()->yes()) {
```

Comment:
While I do like early exists, I'm not the biggest fan of this which the code-style kind of forced me too. Let me know if you have any better ideas, I wanted to avoid having to loop over the arrays again just for that. But it would read nicer for sure :)